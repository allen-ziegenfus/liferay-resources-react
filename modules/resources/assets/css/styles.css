.resources-wrapper {
	text-align: center;
}

.featured .resource-wrapper {
	background-color: #134194;
	color: white;
	height: 100%;
}

.resource.has-resource-badge {
	position: relative;
}

.resource.has-resource-badge .resource-badge {
	background: #eef2fa;
	color: #134194;
	display: inline;
	font-size: 14px;
	left: 1em;
	padding: 1em;
	position: absolute;
	text-transform: uppercase;
	top: 1em;
	z-index: 10;
}

.small-caps {
	font-size: 0.833125em;
	font-weight: 400;
	letter-spacing: 1px;
	text-transform: uppercase;
}

.hide {
	display: none;
}

.w50 {
	width: 50%;
}

.vocabulary-categories-section {
	justify-content: center;
	margin: 0 auto;
	max-width: 1200px;
}

.filter-button {
	margin-left: 2em;
}

.asset-entry.resource a {
	border-radius: 4px;
	box-shadow: 0 0 0.625em 0 rgba(100, 116, 128, 0.16),
		0 0.25em 0.5em -0.375em rgba(100, 116, 128, 0.8);
	color: black;
	display: block;
	overflow: hidden;
	text-decoration: none;
	transition: all 0.2s ease-in-out;
}

.vocabulary-categories-section {
	align-items: center;
}

.asset-entries {
	margin: 0 auto;
	max-width: 1200px;
}

.vocabulary-section.active {
	text-decoration: underline;
}

.vocabulary-see-all a {
	cursor: pointer;
}

.badge-section {
	margin: 2em;
	text-align: center;
}

.featured-resources .resource {
	width: 50%;
}

.resource {
	box-sizing: border-box;
	width: 33%;
}

.standard-padding {
	padding: 1em;
}

.block-container {
	-webkit-box-sizing: border-box;
	box-sizing: border-box;
	display: -webkit-box;
	display: -ms-flexbox;
	display: flex;
	-ms-flex-wrap: wrap;
	flex-wrap: wrap;
	padding: 1em;
}

.page.active {
	font-weight: bold;
}

.page {
	border: 1px solid gray;
	cursor: pointer;
	margin: 0.2em;
	padding: 0.25em;
}

.pager {
	padding: 2em 2em 4em;
	text-align: center;
}

.resources-filter-wrapper {
	margin: auto;
	text-align: left;
}

.applied-category-badge {
	border: 1px solid gray;
	cursor: pointer;
	margin: 0.5em;
	padding: 0.2em 1em 0.2em 0.6em;
	position: relative;
}

.applied-category-badge:after {
	color: gray;
	content: 'x';
	position: absolute;
	right: 0.2em;
	top: 0;
}

.resources-filter-vocabulary {
	margin: 1em;
}

.resources-filter {
	justify-content: center;
	margin: auto;
}

.resources-filter li {
	list-style: none;
}

.resources-filter li input {
	margin-right: 0.5em;
}

.resources-filter button {
	align-self: center;
	margin: 2em;
	max-height: 30px;
	width: auto;
}

.vocabulary-section {
	color: blue;
	cursor: pointer;
	font-size: 2em;
	margin: 2em;
}

.resource {
	min-height: 320px;
}

.resource-wrapper {
	padding: 0;
	position: relative;
	text-align: left;
}

.resource-wrapper img {
	width: 100%;
}

.resource-wrapper .small-caps,
.resource-wrapper .asset-entry-title {
	margin: 0;
	padding: 10px 20px;
}

.resource {
	display: flex;
	flex-direction: column;
}

.resource a {
	flex: 1 0 auto;
}
