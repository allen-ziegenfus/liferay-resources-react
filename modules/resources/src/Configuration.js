let ConfigurationLocal = {
	apiServer: 'http://localhost:8080',
	apiAuth: btoa('test@liferay.com:test'),
	resourceServer: 'https://www-dev.liferay.com'
};

const isLocalhost = Boolean(
	window.location.hostname === 'localhost' ||
		// [::1] is the IPv6 localhost address.
		window.location.hostname === '[::1]' ||
		// 127.0.0.0/8 are considered localhost for IPv4.
		window.location.hostname.match(
			/^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/
		)
);

let Configuration = ConfigurationLocal;

export default Configuration;
