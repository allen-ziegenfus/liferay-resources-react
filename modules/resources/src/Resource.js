import React from 'react';
import Configuration from './Configuration.js';
import Categories from './Categories.js';

let UNDEFINED_VOCBAULARY = 'Other';

function getBadgeText(categories, content) {
	let badgeText = '';
	if (isFeatured(categories)) {
		// TODO: translate this
		badgeText = 'Featured';
	} else if (isNew(content)) {
		badgeText = 'New';
	}

	return badgeText;
}

function getCustomFields(content) {
	let customFields = {};

	if (content.customFields) {
		customFields = content.customFields.reduce(
			(map, field) => ({
				...map,
				[field.name]: field.customValue.data || ''
			}),
			{}
		);
	}
	return customFields;
}

function getCategories(content) {
	let categories = {};
	if (content.taxonomyCategories) {
		categories = content.taxonomyCategories.reduce((map, category) => {
			let property = [
				Categories.categoryToVocab[category.taxonomyCategoryId] ||
					UNDEFINED_VOCBAULARY
			];
			let value = category.taxonomyCategoryName;
			if (map[property]) {
				value = value + ' ' + map[property];
			}

			return {...map, [property]: value || ''};
		}, {});
	}
	return categories;
}

function getUrl(content) {
	let server = Configuration.resourceServer;
	if (content.friendlyUrlPath) {
		return server + '/-/' + content.friendlyUrlPath;
	} else if (content.id) {
		return server + '/d/' + content.id;
	}

	// TODO: error handling if cannot create URL
	return '';
}

function isFeatured(categories) {
	// TODO: Q...Should we be really relying on language here?
	return (
		categories[UNDEFINED_VOCBAULARY] &&
		categories[UNDEFINED_VOCBAULARY].indexOf('Featured') != -1
	);
}

function isNew(content) {
	const dateCreated = Date.parse(content.dateCreated);
	const daysSinceCreated = (new Date() - dateCreated) / (1000 * 3600 * 24);

	return daysSinceCreated < 30;
}

function Resource(props) {
	const resource = props.resource;

	const categories = getCategories(resource.content);
	const customFields = getCustomFields(resource.content);

	const badgeText = getBadgeText(categories, resource.content);

	return (
		<div
			className={
				'asset-entry block resource standard-padding ' +
				(isFeatured(categories) ? 'featured' : '') +
				(badgeText ? ' has-resource-badge' : '')
			}
		>
			<span className="resource-badge">{badgeText}</span>
			<a
				className="element-border font-color no-padding w100"
				href={getUrl(resource.content)}
			>
				<div className="flex flex-column resource-wrapper-recent resource-wrapper">
					<img
						src={
							customFields['idio_small_image'] ||
							Configuration.resourceServer +
								Configuration.defaultResourceImgUrl
						}
					/>
					<div className="small-caps">{categories['Type'] || ''}</div>
					<p className="asset-entry-title">{resource.title}</p>
				</div>
			</a>
		</div>
	);
}

export default Resource;
