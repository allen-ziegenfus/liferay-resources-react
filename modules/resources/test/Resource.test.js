import React from 'react';
import {render} from '@testing-library/react';
import Resource from '../src/Resource';
import Configuration from '../src/Configuration';
import '@testing-library/jest-dom';
import {generateImage} from 'jsdom-screenshot';
import renderer from 'react-test-renderer';
import {toMatchImageSnapshot} from 'jest-image-snapshot';
import '../build/css/styles.css';
expect.extend({toMatchImageSnapshot});

afterEach(() => {
	document.getElementsByTagName('body')[0].innerHTML = '';
});

test('Renders basic web content resource', async () => {
	let resource = {
		content: {
			friendlyUrlPath: 'case-study',
			taxonomyCategories: [
				{
					taxonomyCategoryId: 93352,
					taxonomyCategoryName: 'Case Studies'
				}
			]
		},
		title: 'Test Title'
	};
	const {getByText, queryByText} = render(<Resource resource={resource} />);
	expect(getByText('Test Title')).toBeInTheDocument();
	expect(getByText('Case Studies')).toBeInTheDocument();
	expect(queryByText('Featured')).not.toBeInTheDocument();
	expect(queryByText('New')).not.toBeInTheDocument();
	expect(document.querySelector('a').getAttribute('href')).toBe(
		Configuration.resourceServer + '/-/case-study'
	);
	expect(document.querySelector('img').getAttribute('src')).toBe(
		Configuration.resourceServer + Configuration.defaultResourceImgUrl
	);

	const tree = renderer.create(<Resource resource={resource} />).toJSON();
	expect(tree).toMatchSnapshot();

	const screenshot = await generateImage();
	expect(screenshot).toMatchImageSnapshot();
});

test('Renders featured document resource', async () => {
	const testImageUrl =
		'https://www-dev.liferay.com/documents/10182/244411454/test-image.jpg/41687a03-6722-602f-82e9-21d46f0ad060/test-image.jpg?t=1590459595501';
	let resource = {
		content: {
			dateCreated: new Date(),
			id: 12345,
			customFields: [
				{
					name: 'idio_small_image',
					customValue: {
						data: testImageUrl
					}
				}
			],
			taxonomyCategories: [
				{
					taxonomyCategoryId: 123,
					taxonomyCategoryName: 'Featured'
				},
				{
					taxonomyCategoryId: 174033029,
					taxonomyCategoryName: 'Product Info'
				}
			]
		},
		title: 'Test Featured Title'
	};
	const {getByText, queryByText} = render(<Resource resource={resource} />);
	expect(getByText('Test Featured Title')).toBeInTheDocument();
	expect(getByText('Featured')).toBeInTheDocument();
	expect(getByText('Product Info')).toBeInTheDocument();
	expect(queryByText('New')).not.toBeInTheDocument();
	expect(document.querySelector('a').getAttribute('href')).toBe(
		Configuration.resourceServer + '/d/12345'
	);
	expect(document.querySelector('img').getAttribute('src')).toBe(
		testImageUrl
	);

	const tree = renderer.create(<Resource resource={resource} />).toJSON();
	expect(tree).toMatchSnapshot();

	const screenshot = await generateImage();
	expect(screenshot).toMatchImageSnapshot();
});

test('Renders new resource', async () => {
	let resource = {
		content: {
			dateCreated: new Date(),
			taxonomyCategories: [
				{
					taxonomyCategoryId: 93359,
					taxonomyCategoryName: 'Migration Guides'
				}
			]
		},
		title: 'Test New Title'
	};
	const {getByText, queryByText} = render(<Resource resource={resource} />);
	expect(getByText('Test New Title')).toBeInTheDocument();
	expect(queryByText('Featured')).not.toBeInTheDocument();
	expect(getByText('New')).toBeInTheDocument();

	const tree = renderer.create(<Resource resource={resource} />).toJSON();
	expect(tree).toMatchSnapshot();

	const screenshot = await generateImage();
	expect(screenshot).toMatchImageSnapshot();
});
