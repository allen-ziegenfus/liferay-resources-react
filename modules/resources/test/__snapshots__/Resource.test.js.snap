// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders basic web content resource 1`] = `
<div
  className="asset-entry block resource standard-padding "
>
  <span
    className="resource-badge"
  >
    
  </span>
  <a
    className="element-border font-color no-padding w100"
    href="https://www-dev.liferay.com/-/case-study"
  >
    <div
      className="flex flex-column resource-wrapper-recent resource-wrapper"
    >
      <img
        src="https://www-dev.liferay.comundefined"
      />
      <div
        className="small-caps"
      >
        Case Studies
      </div>
      <p
        className="asset-entry-title"
      >
        Test Title
      </p>
    </div>
  </a>
</div>
`;

exports[`Renders featured document resource 1`] = `
<div
  className="asset-entry block resource standard-padding featured has-resource-badge"
>
  <span
    className="resource-badge"
  >
    Featured
  </span>
  <a
    className="element-border font-color no-padding w100"
    href="https://www-dev.liferay.com/d/12345"
  >
    <div
      className="flex flex-column resource-wrapper-recent resource-wrapper"
    >
      <img
        src="https://www-dev.liferay.com/documents/10182/244411454/test-image.jpg/41687a03-6722-602f-82e9-21d46f0ad060/test-image.jpg?t=1590459595501"
      />
      <div
        className="small-caps"
      >
        Product Info
      </div>
      <p
        className="asset-entry-title"
      >
        Test Featured Title
      </p>
    </div>
  </a>
</div>
`;

exports[`Renders new resource 1`] = `
<div
  className="asset-entry block resource standard-padding  has-resource-badge"
>
  <span
    className="resource-badge"
  >
    New
  </span>
  <a
    className="element-border font-color no-padding w100"
    href=""
  >
    <div
      className="flex flex-column resource-wrapper-recent resource-wrapper"
    >
      <img
        src="https://www-dev.liferay.comundefined"
      />
      <div
        className="small-caps"
      >
        Migration Guides
      </div>
      <p
        className="asset-entry-title"
      >
        Test New Title
      </p>
    </div>
  </a>
</div>
`;
